runtime/account_keys_test.go

```diff
@@ -0,0 +1,741 @@
+				PublicKey: publicKey,
+				HashAlgo:  hashAlgo,
+				Weight:    weight,
+				IsRevoked: false,
+			}
+
+			storage.keys = append(storage.keys, accountKey)
+			storage.returnedKey = accountKey
+			return &accountKey, nil
+		},
+
+		getAccountKey: func(address Address, index int) (*AccountKey, error) {
+			if index >= len(storage.keys) {
+				return nil, nil
+			}
+
+			accountKey := storage.keys[index]
+			storage.returnedKey = accountKey
+			return &accountKey, nil
+		},
+
+		removeAccountKey: func(address Address, index int) (*AccountKey, error) {
+			accountKey := storage.keys[index]
+			accountKey.IsRevoked = true
+
+			storage.keys[index] = accountKey
+			storage.returnedKey = accountKey
+
+			return &accountKey, nil
+		},
+
+		emitEvent: func(event cadence.Event) error {
+			storage.events = append(storage.events, event)
+			return nil
+		},
+		decodeArgument: func(b []byte, t cadence.Type) (value cadence.Value, err error) {
+			return json.Decode(b)
+		},
+	}
+}
+
+func addAuthAccountKey(t *testing.T, runtime Runtime, runtimeInterface *testRuntimeInterface) {
+	test := TestCase{
+		name: "Add key",
+		code: `
+				transaction {
+					prepare(signer: AuthAccount) {
+						let key = PublicKey(
+							publicKey: "010203".decodeHex(),
+							signAlgo: SignatureAlgorithm.ECDSA_P256
+						)
+
+						var addedKey: AccountKey = signer.keys.add(
+							publicKey: key,
+							hashAlgo: HashAlgorithm.SHA3_256,
+							weight: 100.0
+						)
+					}
+				}`,
+		args: []cadence.Value{},
+	}
+
+	err := executeTransaction(test, runtime, runtimeInterface)
+	require.NoError(t, err)
+}
+
+func executeTransaction(test TestCase, runtime Runtime, runtimeInterface *testRuntimeInterface) error {
+	args := encodeArgs(test.args)
+	err := runtime.ExecuteTransaction(
+		Script{
+			Source:    []byte(test.code),
+			Arguments: args,
+		},
+		Context{
+			Interface: runtimeInterface,
+			Location:  utils.TestLocation,
+		},
+	)
+	return err
+}
+
+func executeScript(test TestCase, runtime Runtime, runtimeInterface *testRuntimeInterface) (cadence.Value, error) {
+	args := encodeArgs(test.args)
+	value, err := runtime.ExecuteScript(
+		Script{
+			Source:    []byte(test.code),
+			Arguments: args,
+		},
+		Context{
+			Interface: runtimeInterface,
+			Location:  utils.TestLocation,
+		},
+	)
+	return value, err
+}
+
+func encodeArgs(argValues []cadence.Value) [][]byte {
+	args := make([][]byte, len(argValues))
+	for i, arg := range argValues {
+		var err error
+		args[i], err = json.Encode(arg)
+		if err != nil {
+			panic(fmt.Errorf("broken test: invalid argument: %w", err))
+		}
+	}
+	return args
+}
+
+type TestCase struct {
+	name string
+	code string
+	args []cadence.Value
+	keys []AccountKey
+	err  string
+}
+
+func newStorage() *Storage {
+	return &Storage{
+		events: make([]cadence.Event, 0),
+		keys:   make([]AccountKey, 0),
+	}
+}
+
+type Storage struct {
+	events      []cadence.Event
+	keys        []AccountKey
+	returnedKey AccountKey
+}
```

Comment:
Same here, this storage is only used for the account key tests. Maybe rename to avoid confusion in other tests (e.g. this "storage" type doesn't allow storage reads/writes)